### DISCUSSION TITLE
ZLUDA/blog/zludas-third-life/

### DISCUSSION THREAD

**Discussion by giscus** [vosen/zluda_website#1] [Announcements]
# ZLUDA/blog/zludas-third-life/

https://vosen.github.io/ZLUDA/blog/zludas-third-life/

**Comment by MikeLP** [comment]
This sounds amazing! We're all super excited about the new ZLUDA release and thanks for putting in the hard work. The focus on ML and inference is exactly what we've been waiting for!

As someone who develops ML products based on local inference (like chatbots, RAG, etc.) and has made ROCm my go-to platform, I wanted to share some pain points that I'm constantly dealing with and I think ZLUDA could help with.

First off, a lot of devs are still using pre-RDNA1 GPUs - specifically the Radeon VII you mentioned above. This card is a beast because it offers 16GB VRAM and inference performance that's pretty close to newer RDNA cards, all at a relatively low price point and power consumption. I know people who buy like 8-10 of these cards just so they can run bigger models. It's a bummer that most Radeon VII owners won't be able to use ZLUDA.

Another major issue is flash attention. Honestly, it's a total blocker for many awesome libraries if you're using anything from the 6000 (Navi) generation or AMD MI100 (which I'm currently stuck with). And let's be real, most ML enthusiasts can't afford those fancy >40GB VRAM Nvidia cards, and thats why are looking on AMD.

In my perfect world of rainbows and unicorns, it would be amazing if you could just grab a bunch of cheap but fast AMD GPUs for your ML server and just do something like `pip install vllm` without any issues to build your product.

P.S. I used llama3.1 to translate my comment

**Reply by vosen** [reply]
I am sad about Radeon VII too, but we need to prioritize and the cost/benefit ratio just isn't there. Let's hope for contributors.
Flash attention and `vllm` are on our radar. I hope `llm.c` with flash attention will be one of the first things we run

**Reply by mirh** [reply]
I don't know about gfx906, but just for the absolute records at least on linux you should be able to get working rocm [5.7](https://github.com/xuhuisheng/rocm-build/tree/feature/build) and even perhaps [6.1](https://github.com/woodrex83/ROCm-For-RX580/tree/feat/rocm61) even on polaris (as for windows, good luck with directML I guess). Those probably can still do fine for another year or so.

**Comment by mirh** [comment]
I was wondering, any reason for specifically choosing again ROCm rather than oneAPI/hipSYCL?

And even if it wasn't for specifically technical factors (maybe you just feel more experienced with amd's tooling by now, or it's instead that they got some head start with the overall AI ecosystem) I think it would be cool if you could write a post explaining them. All.
Because I do kinda know why "opencl" just is not an equivalent to CUDA, but 95% of people don't (and 99% certainly don't know that ROCm works by compiling and specifying each single architecture at build time).

**Reply by vosen** [reply]
There are many more people with AMD cards than with Intel cards. That's the main factor

**Reply by mirh** [reply]
Clearly. 
But even though I cannot claim to know the amount of hurdles involved, everything I mentioned actually work everywhere (and not in a "potentially some day in the future if somebody will bother to implement a backend" way, at least theoretically SSCP and universal SYCL binaries sound already better at supporting AMD's gpus than AMD's own compiler)

**Reply by vosen** [reply]
SYCL works if you write SYCL _source code_. By necessity, ZLUDA operates on a lovel level and emits _compiler intermediate language_ and uses underlying runtime. And here things are vendor-specific. For AMD it's LLVM bitcode and HIP runtime, for Intel it's SPIR-V (or maybe LLVM bitcode) and Level Zero runtime

**Reply by mirh** [reply]
... and for the latest version of hipSYCL it is LLVM IR that can be (and *is*) [compiled](https://adaptivecpp.github.io/AdaptiveCpp/generic-sscp/) to every architecture?

**Reply by vosen** [reply]
Interesting, something to keep in mind.
Still, PTX instruction set is richer than available SYCL operations and PTX instructions have specific semantics, incompatible with SYCL, that ZLUDA must handle. SYCL is a C++ dialect and can piggyback on clang support. E.g. PTX has defined semantics (clamp) for shifts equal or bigger than the operand bit size. hipSYCL really does not help too much

**Reply by mirh** [reply]
I don't have much more than just a hunch, but it seems incredibly hard for that to be the case even once the plethora of oneapi extensions to the standard are taken into account (considering it can [literally](https://www.phoronix.com/news/AdaptiveCpp-24.06-Released) be faster than native *on nvidia cards*)?
And anyway I thought the "normal mode of operation" of a given api was kinda irrelevant for the purposes here, since you always end up directly translating PTX to LLVM bitcode (which apparently everybody is a fan of nowadays)?

But I guess thanks for the kind of answers I was expecting.

**Comment by papajo-r** [comment]
Since you mentioned Radeon VII please show some love for Vega 64 too + NVME (or other forms of drives like ram drive) expansion for its HBM2 memory (AMD HBCC) >:D 

This would really give new life to Vega64 cards (especially the OCed ones with water cooling) actually maybe make them very desirable machine learning GPUs in terms of performance per usd.

**Comment by gabrielcapilla** [comment]
This project crossed my mind again, I'm glad it's coming back to life! A few months ago I managed to get DaVinci Resolve Studio to run via ZLUDA and detect CUDA on a multi-gpu system of RX 5700 XT and RX 5700 on Arch Linux, although I couldn't get the video to work.

**Comment by FurkanGozukara** [comment]
Amazing news. I hope you get way more successful and make every CUDA depended app works seamlessly on AMD GPUs.

**Comment by Anazark** [comment]
I can't tell you how hyped I am for this, this is the future ! 
I'm just a gamer/AI-lover who doesn't understand coding, but I've been transitioning towards Linux and plan on finally ditching my 3090 for a 6700XT
I was grieving the loss of AI but ROCm and most of all this project rekindled my hope, long live open-source

I really pray for your project to become the Wine equivalent for Cuda and finally free the AI landscape from proprietary libraries
God spede !!

**Reply by mirh** [reply]
That already happened and pretty much any project supports every backend now.

**Comment by zhangdeepseek** [comment]
Wow!Such a great project.I am here waiting for the Intel GPU version,especially for iris Xe.I had to use it :(

**Comment by stradiotto** [comment]
Hi. My name is César, 

I'm using COMFY-UI to generate images / textures for my 3D games, and I'm also interested into producing 3D models

I already use LUMA AI Genie, Hyper 3D and Hunyuan 3D in Hugging Faces, but the problem is that their use are very limited. So I need some solution to produce 3D running locally (Intel 10th gen, 32GB RAM and Arc A770 16GB)

Could you adapt ZLuda also for Intel Arc A series, please? There are four models of image to 3D that I want to use, but always with the message [AssertionError: Torch not compiled with CUDA enabled]

If possible, and yet better, if you teach me how to go around the dependencies from CUDA, so I can use intel for pytorch extensions (to supply this dependency from NVidia GPUs) in python code

Although my machine is very powerful, I depend totally on third parties to pay for them (and the payment is considerably low for me to pay for professionals to help me), and I'm totally dedicated to work in game production. So, I do not have money to pay for your results, like  tech companies are. 

I offer my apologies for my financial limitations blocking me to support your extraordinary initiative.

This is very frustrating.

I'm honored to talk to you. Thanks for your attention

César - Brazil
